runtime/interpreter/value.go

```diff
@@ -15023,6 +15100,97 @@ func NewEnumCaseValue(
 	return v
 }
 
+func (v *CompositeValue) getBaseValue(interpreter *Interpreter, locationRange LocationRange) *EphemeralReferenceValue {
+	return v.base
+}
+
+func (v *CompositeValue) setBaseValue(interpreter *Interpreter, base *CompositeValue) {
+	attachmentType, ok := interpreter.MustConvertStaticToSemaType(v.StaticType(interpreter)).(*sema.CompositeType)
+	if !ok {
+		panic(errors.NewUnreachableError())
+	}
+
+	var baseType sema.Type
+	switch ty := attachmentType.GetBaseType().(type) {
+	case *sema.InterfaceType:
+		baseType, _ = ty.RewriteWithRestrictedTypes()
+	default:
+		baseType = ty
+	}
+
+	// the base reference can only be borrowed with the declared type of the attachment's base
+	v.base = NewEphemeralReferenceValue(interpreter, false, base, baseType)
+}
+
+func (v *CompositeValue) getAttachmentValue(interpreter *Interpreter, locationRange LocationRange, ty sema.Type) Value {
+	return v.GetMember(interpreter, locationRange, attachmentNamePrefix+string(ty.ID()))
+}
+
+func (v *CompositeValue) forEachAttachment(interpreter *Interpreter, locationRange LocationRange, f func(*CompositeValue)) {
+	iterator, err := v.dictionary.Iterator()
+	if err != nil {
+		panic(errors.NewExternalError(err))
+	}
+
+	oldSharedState := interpreter.SharedState.inAttachmentIteration[v]
+	interpreter.SharedState.inAttachmentIteration[v] = true
+	defer func() { interpreter.SharedState.inAttachmentIteration[v] = oldSharedState }()
+
+	for {
+		key, value, err := iterator.Next()
+		if err != nil {
+			panic(errors.NewExternalError(err))
+		}
+		if key == nil {
+			break
+		}
+		if strings.HasPrefix(string(key.(StringAtreeValue)), attachmentNamePrefix) {
+			attachment, ok := MustConvertStoredValue(interpreter, value).(*CompositeValue)
+			if !ok {
+				panic(errors.NewExternalError(err))
+			}
+			f(attachment)
+		}
+	}
+}
+
+func (v *CompositeValue) GetTypeKey(
+	interpreter *Interpreter,
+	locationRange LocationRange,
+	ty sema.Type,
+) Value {
+	attachment := v.getAttachmentValue(interpreter, locationRange, ty)
+	if attachment == nil {
+		return NilValue{}
+	}
+	// dynamically set the attachment's base to this composite
+	attachment.(*CompositeValue).setBaseValue(interpreter, v)
+	return NewSomeValueNonCopying(interpreter, NewEphemeralReferenceValue(interpreter, false, attachment, ty))
+}
+
+func (v *CompositeValue) SetTypeKey(
+	interpreter *Interpreter,
+	locationRange LocationRange,
+	attachmentType sema.Type,
+	attachment Value,
+) {
+	if v.SetMember(interpreter, locationRange, attachmentNamePrefix+string(attachmentType.ID()), attachment) {
```

Comment:
Maybe refactor `attachmentNamePrefix+` into a function. See alsp other uses of generating them member name for the attachment, like `getAttachmentValue` above and `RemoveTypeKey` below